app/agent/block-registry/registry.service.ts

```diff
@@ -1114,14 +1167,236 @@ export class BlockRegistryService {
+      fs.readdir(this.paths.containers).then((x) => x.filter(BlockRegistryService.isContainerId)),
+    ]);
+
+    // Skip any signaled layers, this is needed as otherwise there is a race where we might delete an already loaded image
+    const allDigests: string[] = [];
+    await waitForPromises(
+      allLayers.map(async (layerDigest) => {
+        const stat = await fs.stat(path.join(this.paths.layers, layerDigest, ".loadSignal"));
+        if (stat.nlink === 1) allDigests.push(layerDigest);
+      }),
+    );
+    const referencedManifests = new Set<ImageId>();
+    const referencedContainers = new Set<ContainerId>();
+    const referencedLayers = new Set<string>();
+    const allConfigs = await fs
+      .readdir(this.paths.blockConfig)
+      .then((ids) =>
+        waitForPromises(
+          ids.map((id) =>
+            readJsonFileOfType(path.join(this.paths.blockConfig, id), OBDConfigValidator).catch(
+              () => ({ lowers: [], resultFile: "" } as OBDConfig),
+            ),
+          ),
+        ),
+      );
+    // Calculate the reference counts based on configs. This will also work with container sealing and lazy pulling
+    for (const config of allConfigs) {
+      if (config.hocusImageId) {
+        if (BlockRegistryService.isImageId(config.hocusImageId)) {
+          referencedManifests.add(config.hocusImageId);
+        } else {
+          this.logger.error(`Found invalid image id in block config ${config.hocusImageId}`);
+        }
+      }
+      if (config.upper) {
+        const directoryName = path.basename(path.dirname(config.upper.data));
+        if (BlockRegistryService.isContainerId(directoryName)) {
+          referencedContainers.add(directoryName);
+        } else {
+          this.logger.error(`Found upper layer with invalid name ${directoryName}`);
+        }
+      }
+      for (const lower of config.lowers) {
+        const layerHash = path.basename(path.dirname(lower.file));
+        referencedLayers.add(layerHash);
+      }
+    }
+
+    // Perform garbage collection of dangling resources
+    const tasks = [];
+    for (const manifest of allManifests) {
+      if (referencedManifests.has(manifest as ImageId)) continue;
+      // Ok this manifest wasn't deleted cause removeContent was interrupted
+      tasks.push(fs.unlink(path.join(this.paths.images, manifest)).catch(catchIgnore("ENOENT")));
+    }
+    for (const container of allContainers) {
+      if (referencedContainers.has(container as ContainerId)) continue;
+      // Ok this container wasn't deleted cause removeContent was interrupted
+      tasks.push(
+        fs.rm(path.join(this.paths.containers, container), {
+          recursive: true,
+          force: true,
+        }),
+      );
+    }
+    for (const layerDigest of allDigests) {
+      if (referencedLayers.has(layerDigest)) continue;
+      tasks.push(
+        (async () => {
+          // TODO: analyze whether this actually works or the race is absurdly rare
```

Comment:
if you've never ran this code then it's almost guaranteed to be broken

Reply:
@hugodutka There's a test trying to hit that case, everything seems to be working, but I'm uneasy about the sync mechanism here. The worst which may happen is that GC deletes a layer of a loaded image. Fortunately, with lazy pulling enabled the code will still work as OBD will download the layer again :)